### PR TITLE
Allow to define a custom formatter for logs

### PR DESCRIPTION
**Pull request by ylecuyer**
This PR allows to have a custom formatter for logs through a `log_formatter` dsl inside the configuration file

Example with a new rails app and no `log_formatter`:

```
=> Booting Puma
=> Rails 5.2.3 application starting in development 
=> Run `rails server -h` for more startup options
Puma starting in single mode...
* Version 3.12.1 (ruby 2.5.5-p157), codename: Llamas in Pajamas
* Min threads: 5, max threads: 5
* Environment: development
* Listening on tcp://localhost:3000
Use Ctrl-C to stop
```

With the config :

```
log_formatter do |str|
  "[#{Process.pid}] [#{Socket.gethostname}] #{Time.now}: #{str}"
end
```

```
=> Booting Puma
=> Rails 5.2.3 application starting in development 
=> Run `rails server -h` for more startup options
[25473] [ylecuyer-U32U] 2019-06-10 22:29:43 +0200: Puma starting in single mode...
[25473] [ylecuyer-U32U] 2019-06-10 22:29:43 +0200: * Version 3.12.1 (ruby 2.5.5-p157), codename: Llamas in Pajamas
[25473] [ylecuyer-U32U] 2019-06-10 22:29:43 +0200: * Min threads: 5, max threads: 5
[25473] [ylecuyer-U32U] 2019-06-10 22:29:43 +0200: * Environment: development
[25473] [ylecuyer-U32U] 2019-06-10 22:29:43 +0200: * Listening on tcp://localhost:3000
[25473] [ylecuyer-U32U] 2019-06-10 22:29:43 +0200: Use Ctrl-C to stop
```

### CODE REVIEW

**Comment by ylecuyer** [comment]
@nateberkopec can't this PR be included in the incoming 4.0.1 release ?

**Comment by nateberkopec** [comment]
That wouldn't be very SemVer of me, would it? 😆

**Comment by nateberkopec** [comment]
Thanks for keeping this updated, I think we can get it in w/4.1.

**Comment by ylecuyer** [comment]
OK great to hear thanks

**Comment by nateberkopec** [comment]
I'm cool with this. Needs a test.

**Comment by nateberkopec** [comment]
Still needs a test here if it will get into 4.1

**Comment by ylecuyer** [comment]
I have added some tests

File: lib/puma/launcher.rb [hunk 1]
```diff
@@ -81,7 +83,7 @@ def initialize(conf, launcher_args={})
       set_rack_environment
 
       if clustered?
-        @events.formatter = Events::PidFormatter.new
+        @events.formatter = Events::PidFormatter.new unless @options[:log_formatter]
```
Comment (nateberkopec) [inline lib/puma/launcher.rb]: Actually, let's just make this `@events.formatter ||= Events::PidFormatter.new`. I don't like have two conditionals looking at `options[:log_formatter]`.

File: History.md [hunk 2]
```diff
@@ -11,8 +11,7 @@
 * 2 bugfixes
   * Fix socket removed after reload - should fix problems with systemd socket activation. (#1829)
   * Add extconf tests for DTLS_method & TLS_server_method, use in minissl.rb. Should fix "undefined symbol: DTLS_method" when compiling against old OpenSSL versions. (#1832)
-* 1 other
```
Comment (nateberkopec) [inline History.md]: ? Why remove?

File: test/test_events.rb [hunk 3]
```diff
@@ -152,7 +152,7 @@ def test_pid_formatter
   end
 
   def test_custom_log_formatter
```
Comment (nateberkopec) [inline test/test_events.rb]: Oh, I didn't even realize this is a thing.

So what does this PR give us? I'm not sure the CustomFormatter class does anything, as it has the same interface as a Proc.

However, I think the config option is useful. I'm not sure how one would have added a custom formatter before 

I'm not sure there's any benefit to the CustomFormatter class added here.

File: test/test_cli.rb [hunk 4]
```diff
@@ -253,6 +253,30 @@ def test_state_file_callback_filtering
...
+    assert_instance_of Puma::Events::PidFormatter, cli.launcher.events.formatter
+  end
+
+  def test_log_formatter_custom_single
+    cli = Puma::CLI.new [ "--config", "test/config/custom_log_formatter.rb" ]
+    assert_instance_of Puma::Events::CustomFormatter, cli.launcher.events.formatter
+  end
+
+  def test_log_formatter_custom_clustered
+    skip_on :jruby, :windows
```
Comment (nateberkopec) [inline test/test_cli.rb]: Would prefer @MSP-Greg's approach to the skips here.

File: lib/puma/launcher.rb [hunk 5]
```diff
@@ -63,6 +60,11 @@ def initialize(conf, launcher_args={})
       @options = @config.options
       @config.clamp
 
+      @events.formatter = Events::CustomFormatter.new(@options[:log_formatter]) if @options[:log_formatter]
+
+      @binder        = Binder.new(@events)
```
Comment (nateberkopec) [inline lib/puma/launcher.rb]: Pretty sure moving this is unnecessary, let's keep it where it is.

File: lib/puma/launcher.rb (hunk 1 above)
Comment (nateberkopec) [reply lib/puma/launcher.rb]: Oh right, duh, there's already a formatter.

Instead, let's just do it all in one conditional then. Keep our impact in as much of one place as possible.

```ruby
@events.formatter = if options[:log_formatter]
  options[:log_formatter]
elsif clustered?
  Events::PidFormatter
end
```

File: History.md (hunk 2 above)
Comment (ylecuyer) [reply History.md]: This is an error I'll revert

File: test/test_cli.rb (hunk 4 above)
Comment (ylecuyer) [reply test/test_cli.rb]: done :+1:

File: test/test_events.rb (hunk 3 above)
Comment (ylecuyer) [reply test/test_events.rb]: i remove the CustomFormatter class and used a proc

File: lib/puma/launcher.rb (hunk 1 above)
Comment (ylecuyer) [reply lib/puma/launcher.rb]: Did something a little different so I don't have to add an else case with the default formatter

**Pull request merged**